### PR TITLE
Improve performance for binning few events into many bins

### LINKED ISSUE #3130: Investigate `sc.group` baseline cost scaling with the number of groups
image

We expect the term that is "constant" in `nevent`, but it is not clear why it is quite so high. Check if all suboperations are multi-threaded, etc.

```python
import scipp as sc

def make_events(ndet, nevent):
    table = sc.data.table_xyz(nevent)
    table.coords['pix'] = (table.coords.pop('z') * ndet).to(dtype='int64')
    table.coords['pix'].unit = None
    return table

pix = sc.arange('pix', ndet, unit=None)
events = make_events(ndet=ndet, nevent=nevent)
events.group(pix)
```

### PR DESCRIPTION
**Pull request by SimonHeybrock**
Fixes #3130 (there are definitely more potential gains to be made, but for now this seems decent).

There are three independent changes which together result in the following speedup:

image

We can see a significant speedup in all relevant cases. There are some minor slowdowns for few detectors, but thus may be due to noise, and should be irrelevant on practice, especially when considering the absolute timings after this change:

image

### CODE REVIEW

File: lib/core/test/subbin_sizes_test.cpp [hunk 1]
```diff
@@ -62,16 +62,12 @@ TEST_F(SubbinSizesTest, sum) {
...
-  x.trim_to({1, {6, 6, 6, 6, 6}});
-  EXPECT_EQ(x, SubbinSizes(1, {0, 0, 2, 3, 0}));
-  x.trim_to({2, {6, 6, 6}});
-  EXPECT_EQ(x, SubbinSizes(2, {0, 2, 3}));
+TEST_F(SubbinSizesTest, exclusive_scan) {
+  SubbinSizes accum(1, {1, 2, 3});
+  SubbinSizes x(2, {2, 0, 3});
+  accum.exclusive_scan(x);
+  EXPECT_EQ(accum, SubbinSizes(2, {2 + 2, 3 + 0, 0 + 3}));
+  EXPECT_EQ(x, SubbinSizes(2, {2, 3, 0}));
```
Comment (jl-wynen) [inline lib/core/test/subbin_sizes_test.cpp:70]: `exclusive_scan` has a bunch of edge cases (like `x` or `this` being empty). Can you dd tests for those?

Also, what happens if `x` and `accum` ave different numbers of elements? can this even happen?

File: lib/dataset/bins.cpp [hunk 2]
```diff
@@ -215,8 +215,12 @@ Variable lookup_previous(const DataArray &function, const Variable &x, Dim dim,
 Variable pretend_bins_for_threading(const DataArray &da, Dim bin_dim) {
   const auto dim = da.dims().inner();
   const auto size = std::max(scipp::index(1), da.dims()[dim]);
-  // TODO automatic setup with reasonable bin count
-  const auto stride = std::max(scipp::index(1), size / 24);
+  const auto nthread = size > 10000000  ? 24
```
Comment (jl-wynen) [inline lib/dataset/bins.cpp:218]: Is this the number of tbb threads? I don't see how you can know the number of threads here.

**Review by jl-wynen — COMMENTED** [review]
Did you benchmark the changes individually? I'm curious how relevant they each are.

**Comment by SimonHeybrock** [comment]
> Did you benchmark the changes individually? I'm curious how relevant they each are.

The threading was the largest contribution, see figures in https://github.com/scipp/scipp/issues/3130#issuecomment-1519394068. The other two I did not record individually, but both were significant.

File: lib/dataset/bins.cpp (hunk 2 above)
Comment (SimonHeybrock) [reply lib/dataset/bins.cpp:218]: True, it is not the number of threads: It is the number of input chunks that is created, which will give the maximum number of threads that could operate on the data.

File: lib/core/test/subbin_sizes_test.cpp (hunk 1 above)
Comment (SimonHeybrock) [reply lib/core/test/subbin_sizes_test.cpp:70]: > Also, what happens if x and accum ave different numbers of elements? can this even happen?

Not unless we have a bug in our `cumsum`/`transform` implementation.
